instrumentation/opentelemetry-instrumentation-openai/src/opentelemetry/instrumentation/openai/patch.py

```diff
@@ -0,0 +1,438 @@
+                llm_prompts.append(tool_calls)
+            else:
+                llm_prompts.append(item)
+
+        span_attributes = {
+            **get_llm_request_attributes(kwargs, prompts=llm_prompts),
+        }
+
+        attributes = LLMSpanAttributes(**span_attributes)
+
+        span = tracer.start_span(
+            "openai.completion",
+            kind=SpanKind.CLIENT,
+            context=set_span_in_context(trace.get_current_span()),
+        )
+        _set_input_attributes(span, kwargs, attributes)
+
+        try:
+            result = wrapped(*args, **kwargs)
+            if is_streaming(kwargs):
+                prompt_tokens = 0
+                for message in kwargs.get("messages", {}):
+                    prompt_tokens += calculate_prompt_tokens(
+                        json.dumps(str(message)), kwargs.get("model")
+                    )
+
+                if (
+                    kwargs.get("functions") is not None
+                    and kwargs.get("functions") != NOT_GIVEN
+                ):
+                    for function in kwargs.get("functions"):
+                        prompt_tokens += calculate_prompt_tokens(
+                            json.dumps(function), kwargs.get("model")
+                        )
+
+                return StreamWrapper(
+                    result,
+                    span,
+                    prompt_tokens,
+                    function_call=kwargs.get("functions") is not None,
+                    tool_calls=kwargs.get("tools") is not None,
+                )
+            else:
+                _set_response_attributes(span, kwargs, result)
+                span.set_status(StatusCode.OK)
+                span.end()
+                return result
+
+        except Exception as error:
+            span.record_exception(error)
+            span.set_status(Status(StatusCode.ERROR, str(error)))
+            span.end()
+            raise
+
+    return traced_method
+
+
+def get_tool_calls(item):
+    if isinstance(item, dict):
+        if "tool_calls" in item and item["tool_calls"] is not None:
+            return item["tool_calls"]
+        return None
+
+    else:
+        if hasattr(item, "tool_calls") and item.tool_calls is not None:
+            return item.tool_calls
+        return None
+
+
+@silently_fail
+def _set_input_attributes(span, kwargs, attributes):
+    tools = []
+    for field, value in attributes.model_dump(by_alias=True).items():
+        set_span_attribute(span, field, value)
+
+    if kwargs.get("functions") is not None and kwargs.get("functions") != NOT_GIVEN:
+        for function in kwargs.get("functions"):
+            tools.append(json.dumps({"type": "function", "function": function}))
+
+    if kwargs.get("tools") is not None and kwargs.get("tools") != NOT_GIVEN:
+        tools.append(json.dumps(kwargs.get("tools")))
+
+    if tools:
+        set_span_attribute(span, SpanAttributes.LLM_TOOLS, json.dumps(tools))
```

Comment:
this is not covered by the otel semconv.

In the current spec prompts/completions should be recorded as a whole on span events - see https://github.com/open-telemetry/semantic-conventions/blob/v1.27.0/docs/gen-ai/gen-ai-spans.md#events

Reply:
agreed, will discard this attribute and merge it inside `gen_ai.prompt`